### ISSUE TITLE
Execute JavaScript in Headless Mode

### ISSUE THREAD

**Issue by captaindavepdx** [gildas-lormeau/SingleFile#196]
**Is your feature request related to a problem? Please describe.**
No. 

**Describe the solution you'd like**
The ability to scroll before downloading a web page in Command Line Interface. This is available in the Chrome, Firefox extension. It'd be great to be able to execute JavaScript in the headless browser before a web page is downloaded.

**Describe alternatives you've considered (optional)**
Use AppleScript to execute javascript in the browser to scroll and click "Single-File" extension.
Learn Casperjs, Phantomjs to use a headless browser to scroll and download a website.
- Status: Learning how to do both.

**Additional context (optional)**
This feature would be greatly appreciated. If it's not part of your product roadmap, then it'd be incredibly helpful to know how to effectively capture all photos on website. This feature will be used as an internal tool. It will absolutely not be published, monetized.

**Comment by gildas-lormeau** [comment]
Actually, there is a simple solution to this issue: set the height of the browser to a big value like 10000px with the `--browser-height` switch. Do you think this solution would be sufficient for you?

**Comment by gildas-lormeau** [comment]
However I agree this feature would be nice to have in order to handle special use cases.

**Comment by gildas-lormeau** [comment]
I added an option "--browser-scripts" allowing to pass a list of script paths relative to the "cli" directory separated by a space. These scripts are executed before any SingleFile's scripts.

**Comment by gildas-lormeau** [comment]
Example with `autoscroll.js` in the `cli` directory:
`./single-file "https://www.example.com" --browser-scripts="./autoscroll.js"`

**Comment by gildas-lormeau** [comment]
I also fixed a regression preventing the loading of deferred resources to work properly in headless mode.

**Comment by captaindavepdx** [comment]
That's funny, there seemed to be a new item in the 'help' menu.

A troubleshooting issue came up since your first reply. Command Line Interface stated 'docHelper' could not be found.

Searched online, reviewed js line by line, and finally realized it was caused by installing 'puppeteer' in the wrong order - 'npm install' then 'npm i puppeteer'

After uninstalling, and reinstalling in the correct order, it worked - 'npm i puppeteer' then 'npm install.'

Thank you, your quick response this morning. Just tested your initial suggestion - it didn't work. It downloaded the same length of web page as before. 

Also, the other video images appear to have changed - now displays a message 'We cannot play the video in this browser. Please try a different web browser.' 

Going to try your new script update now, thank you for being so helpful, fast, and resourceful!

**Comment by gildas-lormeau** [comment]
@captaindavepdx You're welcome. Please do a simple test without passing any particular option and let me know if it works now.

**Comment by gildas-lormeau** [comment]
The fact that the trick with `browser-height` did not work surprises me. I'll do some tests when I can.

**Comment by gildas-lormeau** [comment]
I reproduce the issue with the media that can't be played. That's weird,..

**Comment by captaindavepdx** [comment]
Nice! Happy to help test! Are you wanting to test a large height value, with all other options set to default?

**Comment by gildas-lormeau** [comment]
You would need to verify the switch `--browser-height` works as expected by launching SingleFile in debug mode (`--browser-debug=true`). Thus, it won't run headless. You can do this test if you want :)

**Comment by gildas-lormeau** [comment]
Regarding the issue, it would be great if you could write a script and verify it's running as you expect with the new `--browser-scripts` switch. Note that the script is run very early, before `document.body` even exists in the DOM. So, you may have to register an event handler listening the `load` event dispatched by the `window` if you need to access to this property.

**Comment by captaindavepdx** [comment]
@gildas-lormeau 

Tried a few variations:

**1. --browser-height 3000 ---browser-debug = true**

- Chromium launched, manual scroll was unresponsive, downloaded regular web page.
- Images that were missing are visible, images of videos still replaced by error message.

**2. --broswer-height 30000 --browser-debug = true**
- Same result as test # 1.

**3. ---browser-debug = true**
- Chromium loaded, manual scroll responsive, didn't scroll, downloaded regular web page.
- Images that were missing are visible, some images missing, images of videos visible.
- Suspected missing images caused by default max idle time.
- Increased max idle time, some images still missing.

**4. ---browser-debug = true**
- Chromium loaded, manual scroll responsive, scrolled, downloaded extended web page.
- Images that were missing are visible, some images missing, images of videos visible.
- Suspected missing images caused by default max idle time.
- Increased max idle time, some images still missing, broken layout.

**Comment by captaindavepdx** [comment]
Good to know, will do. Thanks!

**Comment by captaindavepdx** [comment]
@gildas-lormeau 

Threw together a test script - display an alert when the window is finished loading.

https://pastebin.com/Dg3pYvHy

**Tested it on Command Line Interface:**  
--browser-debug = true 
--load-deferred-images-max-idle-time 6000 
--browser-"./test.js

- It ran 8 times, didn't complete.
- Displayed an alert 8 times.
- Printed to console 0 times.

<img width="897" alt="Screen Shot 2019-04-26 at 5 49 57 PM" src="https://user-images.githubusercontent.com/30560737/56842653-75858a80-684c-11e9-8b62-ed7181fa6c1a.png">

**Comment by gildas-lormeau** [comment]
@captaindavepdx Thanks for the tests! Regarding the browser height, I did the same kind of tests and I confirm the `--browser-height` switch did not work as expected. It's fixed. Now, when I run the command below, the whole page is saved into `tmp.html`. Please confirm it's okay on your end too.
`./single-file "https://twitter.com/mcuban" tmp.html --browser-height=30000 --load-deferred-images=false`

Regarding the script injection. I cannot reproduce your issue. On my end, your script runs fine. Do you confirm the `test.js` file is in the `cli` directory? Is your code of SingleFile up to date?

**Comment by captaindavepdx** [comment]
@gildas-lormeau Of course! Happy to help!

What's the steps to update SingleFile?

**Comment by gildas-lormeau** [comment]
Good question! Uninsntall (i.e. delete all files) and install should wor. Otherwise, for minor updates, you can save the `node_modules` folder somewhere. Delete everything, unzip the master.zip file and move the `node_modules` folder.

**Comment by captaindavepdx** [comment]
@gildas-lormeau K, thanks! Is the master zip file up-to-date, and is this the correct link?

https://github.com/gildas-lormeau/SingleFile/archive/master.zip

**Update:**

Uninstalled, reinstalled SingleFile. Tested for height and scripts.

**--browser-height 30000 --load-deferred-images  = false:**
- Downloaded a regular webpage, all regular images worked, video images displayed error message.

**--browser-scripts ./test.js:**
- Same error. Displayed about 8 times, didn't complete, printed 0 times to console.
- Also, tried more complex script - it executed in chrome, crashed cli, no website downloaded.

<img width="1280" alt="Screen Shot 2019-04-27 at 1 07 01 PM" src="https://user-images.githubusercontent.com/30560737/56854558-6efb1f80-68ed-11e9-927e-7a96da3d707a.png">

**Comment by gildas-lormeau** [comment]
The syntax is invalid, but I'm not sure... Some "=" are missing. Please run exactly this to do the test:

`sudo ./single-file "https://twitter.com/mcuban" tmp.html --browser-height=30000 --load-deferred-images=false`

**Comment by captaindavepdx** [comment]
**Update:**
Didn't realize you needed it ran exact. Copied, pasted your code in Command Line Interface.

sudo ./single-file "https://twitter.com/mcuban" tmp.html --browser-height=30000 --load-deferred-images=false
- Same outcome, downloaded a regular web page, all regular images worked, video images displayed an error message. 

Going to uninstall, reinstall again. Want to double check the steps needed to update.

**Download link:**
https://github.com/gildas-lormeau/SingleFile/archive/master.zip

**Uninstall steps:**
Delete SingleFile folder, including node_modules.

**Reinstall steps:**
Unzip file, install puppeteer, install npm, update executable path.

**Comment by gildas-lormeau** [comment]
I confirm it's ok but you would not need to re-install puppeteer normally. Please re-install it anyway since this may be the cause of your current issue. Note that I never saw your error in the past. It looks like puppeteer cannot connect to chrome.

**Comment by gildas-lormeau** [comment]
The good news is that the saved page is fine so the trick with the `--browser-height=30000` option (emulating a very big screen) seems to be effective for loading images.

**Comment by captaindavepdx** [comment]
@gildas-lormeau 

Sent update, it posted before your latest messages.

**Comment by gildas-lormeau** [comment]
Unfortunately I don't have a Mac. I can do tests on Windows and Linux only. All I can say is that none of your additional info are familiar to me. I'm surprised it's so complicated to run Puppeteer on Mac and that you need to be root. Maybe you should use Selenium instead of Puppeteer. Do you confirm that this is a regression and this error now appears regardless of the option you pass to SingleFile?

**Comment by gildas-lormeau** [comment]
By the way, did you try to use the option `--browser-executable-path` to pass the the path of Chrome executable?

**Comment by captaindavepdx** [comment]
Can try Selenium, those solutions took about half a day to troubleshoot - not easily found online.

Just finished documenting solutions for troubleshooting on mac. Hopefully, it helps your mac users.

Still need to uninstall, reinstall again, try your new suggestions.

**Comment by gildas-lormeau** [comment]
Actually, you should just follow the instructions. If it does not work then it's an issue. I mean that there is no troubleshooting documentation to write. I can add the needed info in the README.MD file.

However, I "need" to close this one because from my point of view, the issue "Execute JavaScript in Headless Mode" is fixed and none of the other issues are not related to this one. If may be wrong though.

The other problem with the troubleshooting guide is that it will be almost obsolete since you tested code that changed. For example the "docHelper" issue is a regression I fixed and you just had to update the code. This issue will never appear anymore in the future. I would also like to know *why*, for example, you run `sudo npm i puppeteer --unsafe-perm` (and all the other hacks one by one)? What kind of error appear if you don't run this command? Finally, can we discuss about each issues (because you seem to have a lot of issues actually and I'm a bit lost) separately?

**Comment by captaindavepdx** [comment]
That's fine, I haven't posted to GitHub before. This is my first post, so if it's wrong, please excuse my mistake - only trying to help.

I tried to follow your instructions. But, as a new developer, it wasn't easy. Things that are clear in hindsight, were unclear before. 

For example on Mac, I assumed the executable file was the "Chrome" app icon - like on Windows. That led to an EACCES error. Found out, you had to "right-click" on app icon, then click on "Show Package Contents." Then navigate to /Contents/MacOS/Chromium. And that file is the executable file on Mac, which I've never done that before, didn't know it was there.

<img width="567" alt="Screen Shot 2019-04-27 at 4 23 27 PM" src="https://user-images.githubusercontent.com/30560737/56856122-d4a8d500-6908-11e9-9a78-52f42322ce78.png">

That wasn't mentioned in the instructions. At the time, there were multiple problems happening at once, and I didn't understand it.

The things I posted were things - off the top of my head - that helped get SingleFile to work. I took the time to write it to help your Mac users avoid that frustrating process - I didn't realize it wasn't useful or why it wasn't helpful.

Also, I realize that some of my solutions were unnecessary. For example, you're right, I retried "--browser-executable-path" in Command Line Interface, using the extended path I mentioned above, and it worked. I also retried it using the path to the app icon, which caused the error in the photo above.

So, I tried to follow the instructions, but it didn't work. And there weren't docs, or previous posts about the errors I had in your Github. And the problems came up, one after the other - probably all related to the executable path - but I didn't know that then. 

I also didn't take the time to reproduce the errors before writing. 

To summarize, I've taken a good mental note, and I'll be a better GitHub contributor thanks to your feedback. So, thank you!

Now, I'm going to uninstall, reinstall, and test the execute javascript feature. I'll let you know how that specific problem goes.

**Comment by gildas-lormeau** [comment]
You're welcome. I want also to thank you since you helped me to identify and fix 2 issues with very clear and detailed info. Your bug reports are fine. Feel free to re-open this one if your tests show the current issue is not fixed.

I really want to make things easier for users and fix issues. I know the README.MD in the `cli` directory is a bit light. It's almost deliberate actually because I consider that if someone uses the CLI tool, there are great chances he's a pro and he's doing B2B, and he should maybe contact me for the license... I did not anticipate it could be difficult for other users to identify where the executable of Chrome is located or to know how to use the quotes when special characters are present. I think another manual should be written that would be more detailed, for each major OSes, with screenshots etc. It just takes time and my priority is the `extension` and the `lib` code.

**Comment by captaindavepdx** [comment]
@gildas-lormeau Thanks, happy to help!

I've uninstalled, reinstalled your files, puppeteer, and npm at least a dozen times - with and without extra hacky code. 

I'm not sure why, but this time, puppeteer and npm installed without sudo or --unsafe-perm.

Whatever the reason, that seemed good!

I tested the specific code you recommended - with the addition of "--browser-executable-path"

I also tested a few other option combinations - separately. And I noticed something interesting:

All height settings maxed out at around 8k. And it's the only setting that made a difference, so far.

**./single-file "https://twitter.com/mcuban" tmp.html --browser-height=30000 --load-deferred-images=false**
Downloaded the regular website, height maxed out at around 8k, all of the images worked.

**./single-file "https://twitter.com/mcuban" tmp.html --browser-height=3000 --load-deferred-images=false**
Downloaded the regular website, height maxed out at around 8k, several images missing.

**./single-file "https://twitter.com/mcuban" tmp.html --load-deferred-images=false**
Downloaded the regular website, height maxed out at around 8k, many images missing.

**./single-file "https://twitter.com/mcuban" tmp.html**
Downloaded the regular website, height maxed out at around 8k, many images missing.

**Comment by captaindavepdx** [comment]
@gildas-lormeau Also, tried to add the "--browser-scripts" to a basic run with minimal options.

**./single-file "https://twitter.com/mcuban" tmp.html --browser-executable-path="/Applications/Google Chrome.app/Contents/MacOS/Google Chrome" --browser-scripts="./test.js"**
Didn't complete, displayed an error, "Error: Evaluation failed: ReferenceError: docHelper is not defined"

**Comment by captaindavepdx** [comment]
@gildas-lormeau Also, my script will cause the window to redirect to a new url in the same window. So, it will lose the script when it changes url.

It looked like you made it possible to add multiple scripts in cli. Is there a way to load a specific script for one url, and load another script for another url?

**Comment by gildas-lormeau** [comment]
@captaindavepdx Please create a separate issue for the new "8k" issue. I cannot reproduce both of the issues you're describing. The second one could be related to an invalid syntax into test.js if the error only appears when you add `--browser-scripts="./test.js"`. Your last question can be solved with a test on the `location.href`, I guess. Again it's a separate issue...

**Comment by captaindavepdx** [comment]
K, will do, thanks!

**Comment by captaindavepdx** [comment]
Will do, thanks!

**Comment by gildas-lormeau** [comment]
I'm glad to know you confirm this issue is fixed :)

**Comment by captaindavepdx** [comment]
Also, found a way to stop the script(s) from firing repeatedly on websites (Mac)
Apparently, if you add an if-then statement, and set it to fire based on thay url, it only fires once. If you use multiple scripts, they all must have an if-then statement that fires based on each script's url.
